File: algorithm study/BOJ/2. Silver/11660.cpp
// boj 11660 구간 합 구하기 5
// prefix sum
#include <iostream>
using namespace std;

long long N, M;  // 1 ≤ N ≤ 1024, 1 ≤ M ≤ 100,000
int arr[1025][1025];
long long temp[1025][1025];  // 누적합

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> N >> M;
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= N; j++) {
            int num;
            cin >> num;
            arr[i][j] = num;
        }
    }

    for (int i = 1; i <= N; i++) temp[i][1] = arr[i][1];  // 행별로 누적합
    for (int i = 1; i <= N; i++) {
        for (int j = 2; j <= N; j++) {
            temp[i][j] = temp[i][j - 1] + arr[i][j];
        }
    }

    for (int i = 0; i < M; i++) {
        long long sum = 0;
        int x1, y1, x2, y2;  // x1 ≤ x2, y1 ≤ y2
        cin >> x1 >> y1 >> x2 >> y2;
        if (x1 == x2 & y1 == y2)
            sum = arr[x1][y1];
        else {
            for (int a = x1; a <= x2; a++) {
                sum += temp[a][y2] - temp[a][y1 - 1];
            }
        }
        cout << sum << "\n";
    }

    return 0;
}

File: algorithm study/BOJ/2. Silver/11660.js
// boj 11660 구간 합 구하기 5
// prefix sum

const input = require("fs").readFileSync("./example.txt").toString().split("\n");
// const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

const [n, m] = input[0].split(" ").map(Number);
let arr = [];
for (let i = 1; i <= n; i++) {
  arr.push(input[i].split(" ").map(Number));
}

let prefixsum = Array.from(Array(n + 1), () => Array(n + 1).fill(0));
for (let i = 1; i <= n; i++) {
  for (let j = 1; j <= n; j++) {
    if ((i === j) === 1) prefixsum[i][j] = arr[i][j];
    else {
      prefixsum[i][j] = prefixsum[i - 1][j] + prefixsum[i][j - 1] - prefixsum[i - 1][j - 1] + arr[i - 1][j - 1];
    }
  }
}

let result = "";
for (let i = 0; i < m; i++) {
  const [x1, y1, x2, y2] = input[n + 1 + i].split(" ").map(Number);
  let temp = prefixsum[x2][y2] - prefixsum[x2][y1 - 1] - prefixsum[x1 - 1][y2] + prefixsum[x1 - 1][y1 - 1];
  result += temp + "\n";
}
console.log(result);

// const input = require("fs").readFileSync("./example.txt").toString().split("\n");
// // const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

// const [n, m] = input[0].split(" ").map(Number);

// let arr = input.slice(1, n + 1).map((str) => str.split(" ").map(Number));
// let temp = Array.from(Array(n + 1), () => new Array(n + 1).fill(0));

// for (let i = 1; i <= n; i++) temp[i][1] = arr[i - 1][0];
// //행별 누적합이 아닌3 좌표별 누적합
// for (let i = 1; i < n + 1; i++) {
//   for (let j = 1; j < n + 1; j++) {
//     temp[i][j] = arr[i - 1][j - 1] + temp[i][j - 1] + temp[i - 1][j] - temp[i - 1][j - 1];
//   }
// }

// let sum = [];
// for (let i = 0; i < m; i++) {
//   const [x1, y1, x2, y2] = input[n + i + 1].split(" ").map(Number);

//   //겹치는 부분 빼기, 두 번 뺀 부분 더하기
//   sum[i] = temp[x2][y2] - temp[x1 - 1][y2] - temp[x2][y1 - 1] + temp[x1 - 1][y1 - 1];
// }

// for (let i = 0; i < m; i++) console.log(sum[i]);
